Add filter to disable local attachment storage

obenland opened the pull request:
Follow-up to #2314

Addresses https://github.com/Automattic/wordpress-activitypub/pull/2314#discussion_r2472367559

## Proposed changes:
* Adds a new `activitypub_store_attachments_locally` filter that allows users and plugins to disable local storage of attachments from incoming ActivityPub posts
* This addresses the feedback about providing an option for users with limited webspace who may prefer not to download and store remote images locally

### Other information:

- [ ] Have you written new tests for your changes, if applicable?

## Testing instructions:

### Default behavior (no change):
1. Ensure the plugin is active
2. Receive an ActivityPub post with attachments
3. Verify attachments are downloaded and stored locally in `wp-content/uploads/activitypub/ap_posts/{post_id}/`

### Disabled storage:
1. Add this code to your theme's `functions.php`:
```php
add_filter( 'activitypub_store_attachments_locally', '__return_false' );
```
2. Receive an ActivityPub post with attachments
3. Verify that attachments are NOT downloaded or stored locally

### Changelog entry

<!-- Skipping changelog as this is a minor follow-up -->

obenland committed:
Add filter to disable local attachment storage

Introduces the `activitypub_store_attachments_locally` filter to allow
users and plugins to disable local storage of attachments from incoming
ActivityPub posts. This addresses feedback about providing an option for
users with limited webspace who may prefer not to store remote images
locally.

The filter is applied in both the add() and update() methods of the
Posts collection class, with full documentation on the first instance
and a reference comment on the second instance following WordPress
conventions.

Jiwoon-Kim commented:
Some thoughts on caching only the media object and not the media file:  
https://github.com/Automattic/wordpress-activitypub/discussions/2377#discussioncomment-14817500

• **Dual-Layer Persistence:** Create a specialized local data structure — the **Federated Media Object (FMO)** — which stores metadata (e.g., original AS2 ID, URL, and size information) of remote media but does **not** store the related binary file (zero-sideload). This separation of metadata and binary data is decisive for achieving resource optimization.

Yes, of course, this could be a WordPress-specific feature for WP-to-WP media library federation and comment interactions, but I mention it because I think it highlights aspects that could be helpful for improvements.

---

Yes, I proposed the FMO (Federated Media Object) to prevent duplicate caching between the `preview`'s `attachment`, inline images in the `content`, and the `article`'s `attachment` images.

Also, it’s important to note that the image sizes in the `preview`, the `article`, and the original source may each differ in version or resolution.

In this case, specifying the ID of the FMO image object in the `source` property of a Link object could be the most robust approach for federating isolated media objects within the WordPress media library.

obenland committed:
Refactor attachment handling to always delete on update

Refactors the attachment storage logic:
- Extracts filter logic into a reusable maybe_import_attachments() method
- Always deletes existing attachments on update before re-importing
- This ensures the filter return value is respected even if it changed
  since the post was first added

This addresses feedback that the filter value might change between
add and update operations, so we should always clean up and re-evaluate
whether to store attachments locally.

Jiwoon-Kim commented on includes/collection/class-posts.php:
```diff
@@ -132,8 +147,13 @@ public static function update( $activity, $user_id ) {
+
+			if ( $store_locally ) {
+				Attachments::delete_ap_posts_directory( $post_id );
+				Attachments::import_post_files( $activity['object']['attachment'], $post_id );
```
https://github.com/Automattic/wordpress-activitypub/discussions/2382#discussioncomment-14814249
> IV. Garbage Collection and Lifecycle Management
> A major consequence of bypassing the WordPress Media Library is that the standard deletion mechanisms for media are disabled. When the parent CPT (the remote actor record) is deleted, the cached files are left as "orphaned files" on the server, causing disk bloat

pfefferle committed:
Merge branch 'trunk' into add/attachment-storage-filter

matticbot committed:
Prettify

pfefferle commented:
That is not exactly what I had in mind 🤔

I will try to provide an example later today.

Jiwoon-Kim commented:
https://wordpress.org/plugins/regenerate-thumbnails/

Using a plugin like **Regenerate Thumbnails** can pose a significant risk when dealing with a large number of image files. In other words, in the case of **Fediverse**, performing bulk synchronization of images could be very dangerous.

My suggestion is also based on the assumption of using the mediaLib endpoint(+FMO). By specifying the file size in advance, each instance can predictively generate the image itself.

Additionally, retrieving and synchronizing the file URL of remote image objects also becomes easier.

---

Inspired by FileBird, I think that using an S2S Media Library endpoint could make it possible not only to query collections or individual image objects, but even to transfer them in bulk as a compressed archive.

Jiwoon-Kim commented:
The most painful idea I'm dealing with right now is that,  
to implement a **Proxy Post** system, I would need to insert an **actor_id mapping layer** between the `wp_users` table and the `author` field.

However, once `author` is decoupled from `wp_users`,  
the entire **core structure** and **countless plugin/function dependencies** will inevitably break.

I'm trying to design a **blog actor model** that allows multiple intermediaries (proxies or agents) to exist between the author and the post,  
but the architectural complexity is skyrocketing to an unbearable level...

https://github.com/Automattic/wordpress-activitypub/discussions/2322#discussioncomment-14852012

As you know, when enabling Multisite in WordPress, additional tables related to blog_id and site_id are created.

Neither Mastodon nor Misskey supported ActivityPub from the beginning — their architectures had to be restructured to accommodate federation.

So why do you think WordPress could achieve full federation without any core modification?

In my opinion, a more proper and sustainable approach would be to first handle core extensions and housekeeping through a canonical plugin, and only then build the ActivityPub layer on top of that foundation.

obenland commented:
> That is not exactly what I had in mind 🤔 I will try to provide an example later today.

@pfefferle Have you had a chance to give this another thought?

pfefferle committed:
Merge branch 'trunk' into add/attachment-storage-filter
